sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -608,6 +608,12 @@ pi_result piDevicesGet(pi_platform Platform, pi_device_type DeviceType,
   if (NumDevices)
     *NumDevices = ZeDeviceCount;
 
+  if (Devices == nullptr) {
+    assert(NumEntries == 0 &&
+           "NumEntries should be zero when querying the number of devices");
```

Comment:
Actually the spec (https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clGetDeviceIDs.html) says:
>if devices is not NULL, the num_entries must be greater than zero.

I read it that the NumEntries can be anything if Devices is NULL.

Reply:
I don't think so. 
It means if NumEntries is zero, the "devices" parameter must be null.
You can infer this from the following statement in the spec, too.

>  If devices is NULL, this argument is ignored.

Anyway, SYCL RT passes zero for NumEntries and nullptr for devices when querying the number of devices.

Reply:
>If devices is NULL, this argument is ignored.

It is "devices" that is not used them.
There is nowhere it is saying that NumEntries must be zero if Devices is null.
Instead please add an assert that NumEntries >0 if Devices is not null.

> SYCL RT passes zero 

In the long run PI API may be used outside of SYCL, so let's stick to what we said it's semantics is, i.e. follow OpenCL API in this case.

Reply:
It is basic contrapositive rule.
If P-->Q is true, then ~Q-->~P is true.

> if devices is not NULL, the num_entries must be greater than zero.
So, the num_entires is NOT greater than zero, then devices should be null.

> I read it that the NumEntries can be anything if Devices is NULL
You cannot say P-->Q is true, then ~P-->~Q is true.

Reply:
The contrapositive rule allows you to assert that if NumEntries is zero (since it's unsigned) then Devices should be nullptr, but this is not what you are checking. What am I missing?

Reply:
> What am I missing?
Ok. I originally said "It means if NumEntries is zero, the "devices" parameter must be null."
But you did not agree. "There is nowhere it is saying that NumEntries must be zero if Devices is null."
Now you say "The contrapositive rule allows you to assert that if NumEntries is zero (since it's unsigned) then Devices should be nullptr"
I will assume you agreed with my original suggestion, and will change the code accordingly.

Reply:
I didn't agree to your "NumEntries must be zero if Devices is null", not the "It means if NumEntries is zero, the "devices" parameter must be null." Sorry for confusion. Yes, please change to check what is legal to check.

Reply:
No problem. :)
>> The contrapositive rule allows you to assert that if NumEntries is zero (since it's unsigned) then Devices should be nullptr.
I already changed the code based on this statement.

>> Yes, please change to check what is legal to check.
```
  if (NumEntries == 0) {
    assert(Devices == nullptr &&
           "Devices should be nullptr when querying the number of devices");
    return PI_SUCCESS;
  }
```
Isn't this what you said above?

Reply:
indeed.